Requestify regex pattern parsing

hamishknight opened the pull request:
Instead of doing the pattern parsing in both the C++ parser and ASTGen, factor out the parsing into a request that returns the pattern to emit, regex type, and version. This can then be lazily run during type-checking.

hamishknight commented:
@swift-ci please test

hamishknight commented on lib/Sema/TypeCheckRegex.cpp:
```diff
@@ -111,3 +133,62 @@ bool swift::decodeRegexCaptureTypes(ASTContext &ctx,
+  // Let the Swift library parse the contents, returning an error, or null if
+  // successful.
+  size_t version = 0;
+  auto capturesBuf = ctx.AllocateUninitialized<uint8_t>(
+      getCaptureStructureSerializationAllocationSize(regexText.size()));
```
Don't actually have to ASTContext allocate this anymore

hamishknight commented:
@swift-ci please test

DougGregor reviewed (approved):
Looks good, thank you!

hamishknight committed:
NFC: Remove unused var

hamishknight committed:
Requestify regex pattern parsing

Instead of doing the pattern parsing in both the
C++ parser and ASTGen, factor out the parsing into
a request that returns the pattern to emit, regex
type, and version. This can then be lazily run
during type-checking.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test macOS
